[KEYCLOAK-18419] Support SAML 2.0 Encrypted IDs in Assertion

Kanzow-IHK opened the pull request:
The SAML spec 2.0 permits assertions with encrypted IDs, but those are currently not handled by KeyCloak.
This PR enables decryption for encrypted IDs in SAML responses, see [KEYCLOAK-18419](https://issues.redhat.com/browse/KEYCLOAK-18419)

I used the same decryption mechanism that is used to decrypt encrypted assertions (`XMLEncryptionUtil.decryptElementInDocument`)

mhajas reviewed (changes requested):
Thanks for the changes @Kanzow-IHK. It looks good to me, I just added one comment into the unit test.

Also, we should probably include some integration test as well. I believe we could use something similar as in following test:
https://github.com/keycloak/keycloak/blob/a8fdd79d1b9d78e94a7025e2a3f367561fd7557d/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlBrokerTest.java#L303-L322

This test is basically intercepting Saml response from IDP to SP and adding a new `AttributeType` into the assertion. So instead of adding `AttributeType`, we need replacing `baseId` with `encryptedId`. WDYT?

hmlnarik commented:
@Kanzow-IHK Do you plan to continue in this PR by addressing the comments above?

Kanzow-IHK commented:
> @Kanzow-IHK Do you plan to continue in this PR by addressing the comments above?

@hmlnarik Sorry for the late answer, but I have been quite busy, yes I will follow your advice and write two more tests, but I'll need a few days

Kanzow-IHK commented on saml-core/src/test/java/org/keycloak/saml/processing/core/parsers/saml/SAMLParserTest.java:
```diff
@@ -187,6 +186,18 @@ public void testSaml20EncryptedAssertion() throws Exception {
     }
 
+    @Test
+    public void testSaml20EncryptedId() throws Exception {
```
Added Unit Test: https://github.com/keycloak/keycloak/pull/8155/commits/d9e5932745fa3d47a58530eb2960af82374752b0#diff-3c88891cd0d24dd3c0c7bf423ff5d2471782344c848c2496b49973be56f66b9d

Kanzow-IHK replied:
Concerning the integration test, I do not really know how to do that. Can keycloak produce a SAML assertion containing an Encrypted ID?

mhajas reviewed (changes requested):
@Kanzow-IHK We don't need Keycloak to produce an actual assertion with `EncryptedID` for testing this. We can just intercept communication between IDP and consumer and replace `BaseId` with `EncryptedId` directly in sent document. Something [like this](https://github.com/mhajas/keycloak/commit/5f38f81cf0c7b04fdd78f7b448259e0dcf4c1d2e). Feel free to just incorporate my changes into your commit if it makes sense to you. @hmlnarik Could you please check whether the test makes sense?

hmlnarik commented:
> Could you please check whether the test makes sense?

@mhajas Yes, that's the right test, thank you!

Kanzow-IHK commented:
@mhajas  I cherry-picked your suggestion for an integration test and it's working just fine (tested locally). Thank you so much for this, it would've taken me a long time to figure out how to do the test properly!

hmlnarik commented:
@mhajas The quarkus failure seems related, could you please check it?

mhajas commented:
@hmlnarik Sure I will have a look at it

mhajas commented:
I was able to make the test work. However, I was not able to identify the exact cause. The problem seems to be caused by two differences between Quarkus and Undertow serializing/deserializing of SAML messages. 
1. Resulting document on Quarkus doesn't contain duplicate definitions of a namespace, so when we have `xmlns:saml` defined in the Root element, Assertion and possibly NameID (this can be added [here](https://github.com/keycloak/keycloak/blob/master/saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/writers/BaseWriter.java#L268) as last boolean parameter), the resulting document contains only one `xmlns:saml` attribute in the root element.
2. I am not sure about this second thing, but it looks like that when the encrypted element is decrypted, it is decrypted as a separate document that consists of only this one element. However, on Undertow, the deserializer is somehow able to work out that the `saml` prefix corresponds to the assertion namespace even if encrypted `NameId` didn't contain the `saml` prefix definition. This doesn't work on Quarkus.

I was also looking into how it is possible that tests for encryption of Assertions don't suffer from the same issue, and I suspect it is caused by the fact, that we never do the encryption after serializing/deserializing (which means losing `xmlns:saml` definition in case of Quarkus), therefore the namespace definition is always part of the encrypted element.

Anyway, the fix for the test failure is [here](https://github.com/mhajas/keycloak/commit/a090692ccd4ba8fe5c6e9c3afe63b2e95c714534), it simply adds `xmlns:saml` attribute to `NameId` before encryption so that it is known during decryption. @Kanzow-IHK @hmlnarik Could you please let me know whether the fix makes sense to you or we should fix this on the Quarkus side somehow?

hmlnarik commented:
> Anyway, the fix for the test failure is [here](https://github.com/mhajas/keycloak/commit/a090692ccd4ba8fe5c6e9c3afe63b2e95c714534), it simply adds `xmlns:saml` attribute to `NameId` before encryption so that it is known during decryption. @Kanzow-IHK @hmlnarik Could you please let me know whether the fix makes sense to you or we should fix this on the Quarkus side somehow?

Thanks @mhajas. This is sufficient for the test purposes. Proper fix would be necessary if / once we support name ID encryption.

mhajas commented:
Thanks, @hmlnarik! @Kanzow-IHK I will add one more change to the test. With @hmlnarik we agreed that we should make sure that the `NameId` was really encrypted (plain NameID value is not part of the document) and also successfully decrypted by checking whether `nameId` is part of the created user.

Commit:
[KEYCLOAK-18419] Support SAML 2.0 Encrypted IDs in Assertion

mhajas commented:
@Kanzow-IHK There is the last change to the test: https://github.com/mhajas/keycloak/commit/84db785e25e93f3cd5e0cd0433b2c57904aa1d81 could you please cherry-pick both commits?

Kanzow-IHK commented:
> @Kanzow-IHK There is the last change to the test: [mhajas@84db785](https://github.com/mhajas/keycloak/commit/84db785e25e93f3cd5e0cd0433b2c57904aa1d81) could you please cherry-pick both commits?

@mhajas I cherry-picked both your commits

mhajas reviewed (dismissed):
Thank you for your contribution @Kanzow-IHK!

hmlnarik reviewed (dismissed):
Approving based on @mhajas's approval

hmlnarik commented:
@Kanzow-IHK Could you please squash the two commits into a single one? Then this PR would be ready for merging

Kanzow-IHK commented:
> @Kanzow-IHK Could you please squash the two commits into a single one? Then this PR would be ready for merging

@hmlnarik Sorry, I had accidentally merged upstream master into my PR. I do not think we want that, because upstream has already been switched to SNAPSHOT-16.0.0. To fix this, I simply discarded the merge commit, so no more squashing should be needed. I hope that is ok

hmlnarik commented:
> @hmlnarik Sorry, I had accidentally merged upstream master into my PR. I do not think we want that, because upstream has already been switched to SNAPSHOT-16.0.0. To fix this, I simply discarded the merge commit, so no more squashing should be needed. I hope that is ok

Absolutely, thank you
